### ISSUE TITLE
Push Notification issue

### ISSUE THREAD

**Issue by pashtika** [notnoop/java-apns#241]
I am using following code, but not showing any error nor notifications. Please help on this.

```
public class Demo {
	ApnsService pushService = APNS.newService()
			.withCert("/home/home/certificate/ahma_cert.p12", "password")
			.withSandboxDestination()
			.build();

	void notify(String message){
		String payload = APNS.newPayload().alertBody(message).badge(45).sound("default").build();
		String token = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX";
		pushService.push(token, payload);

		Map<String, Date> inactiveDevices = pushService.getInactiveDevices();
		for (String deviceToken : inactiveDevices.keySet()) {
		    Date inactiveAsOf = inactiveDevices.get(deviceToken);
		    System.out.println("Output : " + inactiveAsOf);
		}
	}

	public static void main(String[] args) {
		PushServiceServer object = new PushServiceServer();
		object.notify("Good Morning Poland");
		System.out.println(object);
	}
}
```

**Comment by matzew** [comment]
whats exactly the issue ?

**Comment by pashtika** [comment]
When I run above code, I neither see error message and nor see notifications coming to iOS device. 
1) How can we confirm connection is establish with gateway.sandbox.push.apple.com.
2) How we confirm push notifications being sent to iOS device.

Also, as I said above code not working for me, don't see any errors and notification. What could be the reason of failure?

**Comment by matzew** [comment]
@pashtika you could implement this interface https://github.com/notnoop/java-apns/blob/master/src/main/java/com/notnoop/apns/ApnsDelegate.java

and include it when building the ApnsService e.g.:
```
APNS.newService()
  .withXXXXX().
.....
. withDelegate(new ApnsDelegate() {
/// implement the methods
})
```

**Comment by pashtika** [comment]
How we can send emoji and images in notifications?

**Comment by pashtika** [comment]
Can anyone please reply to above query?

**Comment by pashtika** [comment]
Hi, Could you please reply?

**Comment by harmeetsingh0013** [comment]
@pashtika firstly you need to verify your security certificate generate by mac machine.

**Comment by pashtika** [comment]
yes, OK, I am able to send notification to Apple IOS device. But how can I send emoji and different sound though APNS java code?

**Comment by harmeetsingh0013** [comment]
that's good, @pashtika for this, i have no idea. you will need to explore Java APNS libraries, or use another one. If you did this, please notify us.

**Comment by matzew** [comment]
Perhaps worth to look at Apple's APNs doc? 
https://developer.apple.com/library/ios/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/Chapters/ApplePushService.html

**Comment by pashtika** [comment]
Is there any way to confirm whether notifications being send successfully though APNS java code?

**Comment by matzew** [comment]
to device? no - well, not directly via apns api

to apns itself: yes
(see the ApnsDelegate interface I mentioned on a different thread before)

On Thursday, July 9, 2015, pashtika <notifications@github.com> wrote:

> Is there any way to confirm whether notifications being send successfully
> though APNS java code?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/notnoop/java-apns/issues/241#issuecomment-119924169>.
>


-- 
Sent from Gmail Mobile

**Comment by JavaHelper** [comment]
Could you please share code snippet, how we can use APNSDelegate to know Notifications being sent on iOS device within APNS API?

**Comment by justuno** [comment]
@JavaHelper  you can implement ApnsDelegate's messageSent method. if the message was sent to APNS successfully, this method would be called.this means that the message was accepted by the APNS.

However, it's almost impossible to check that whether the message was arrived at the specified device,because apple doesn't guarantee this.what we only know is that APNS will attempt to send the message,that's it.

**Comment by matzew** [comment]
On Thu, Dec 3, 2015 at 12:39 PM, Justuno <notifications@github.com> wrote:

> @JavaHelper <https://github.com/JavaHelper> you can implement
> ApnsDelegate's messageSent method. if the message was sent to APNS
> successfully, this method would be called.this means that the message was
> accepted by the APNS.
>
> However, it's almost impossible to check that whether the message was
> arrived at the specified device,because apple doesn't guarantee this.what
> we only know is that APNS will attempt to send the message,that's it.
>
Right, but you can build your own layer around (e.g. ping your backend from
the app). That's what we have done for our AeroGear push server:
*
https://aerogear.org/docs/unifiedpush/ups_userguide/index/#_pushapplication_analytics
*
https://aerogear.org/docs/unifiedpush/ups_userguide/index/#_pushapplication_activity



> —
> Reply to this email directly or view it on GitHub
> <https://github.com/notnoop/java-apns/issues/241#issuecomment-161606728>.
>



-- 
Matthias Wessendorf

blog: http://matthiaswessendorf.wordpress.com/
sessions: http://www.slideshare.net/mwessendorf
twitter: http://twitter.com/mwessendorf
